proxylib/binarymemcached/parser.go

```diff
@@ -0,0 +1,364 @@
+func (p *BinaryMemcacheParserFactory) Create(connection *proxylib.Connection) proxylib.Parser {
+	log.Infof("BinaryMemcacheParserFactory: Create: %v", connection)
+	return &BinaryMemcacheParser{connection: connection, injectQueue: make([]queuedInject, 0)}
+}
+
+// compile time check for interface implementation
+var _ proxylib.ParserFactory = &BinaryMemcacheParserFactory{}
+
+var binaryMemcacheParserFactory *BinaryMemcacheParserFactory
+
+const (
+	parserName = "binarymemcache"
+)
+
+func init() {
+	log.Info("init(): Registering binaryMemcacheParserFactory")
+	proxylib.RegisterParserFactory(parserName, binaryMemcacheParserFactory)
+	proxylib.RegisterL7RuleParser(parserName, L7BinaryMemcacheRuleParser)
+}
+
+// BinaryMemcacheParser implements proxylib.Parser
+type BinaryMemcacheParser struct {
+	connection *proxylib.Connection
+
+	requestCount uint32
+	replyCount   uint32
+	injectQueue  []queuedInject
+}
+
+var _ proxylib.Parser = &BinaryMemcacheParser{}
+
+const headerSize = 24
+
+// OnData parses binary memcached data
+func (p *BinaryMemcacheParser) OnData(reply, endStream bool, dataBuffers [][]byte, offset int) (proxylib.OpType, int) {
+	log.Debugf("OnData with offset %d", offset)
+
+	if reply {
+		if p.checkInjectQueue() {
+			return proxylib.INJECT, len(DeniedMsgBase)
+		}
+		if len(dataBuffers) == 0 {
+			return proxylib.NOP, 0
+		}
+	}
+
+	//TODO don't copy data from buffers
+	data := (bytes.Join(dataBuffers, []byte{}))[offset:]
+	log.Debugf("Data length: %d", len(data))
+
+	if headerSize > len(data) {
+		headerMissing := headerSize - len(data)
+		log.Debugf("Did not receive needed header data, need %d more bytes", headerMissing)
+		return proxylib.MORE, headerMissing
+	}
+
+	bodyLength := binary.BigEndian.Uint32(data[8:12])
+
+	keyLength := uint32(binary.BigEndian.Uint16(data[2:4]))
+	extrasLength := uint32(data[4])
+
+	if keyLength > 0 {
+		if int(headerSize+keyLength+extrasLength) > len(data) {
+			keyMissing := int(headerSize+keyLength+extrasLength) - len(data)
+			log.Debugf("Did not receive enough bytes for key, need %d more bytes", keyMissing)
+			return proxylib.MORE, keyMissing
+		}
+	}
+
+	opcode, key, err := p.parse(data, extrasLength, keyLength)
+	if err != 0 {
+		return proxylib.ERROR, int(err)
+	}
+
+	logEntry := &cilium.LogEntry_GenericL7{
+		&cilium.L7LogEntry{
+			Proto: "binarymemcached",
+			Fields: map[string]string{
+				"opcode": strconv.Itoa(int(opcode)),
+				"key":    key,
+			},
+		},
+	}
+
+	// we don't filter reply traffic
+	if reply {
+		log.Debugf("reply, passing %d bytes", len(data))
+		p.connection.Log(cilium.EntryType_Response, logEntry)
+		p.replyCount++
+		return proxylib.PASS, int(bodyLength + headerSize)
+	}
+
+	p.requestCount++
+
+	matches := p.connection.Matches(binaryMemcacheMeta{opcode, key})
+	if matches {
+		p.connection.Log(cilium.EntryType_Request, logEntry)
+		return proxylib.PASS, int(bodyLength + headerSize)
+	}
+
+	magic := 0x81 | data[0]
+
+	if p.requestCount == p.replyCount+1 {
```

Comment:
Add a comment to explain that this guarantees in-order replies

Reply:
done